scientific_library/tvb/analyzers/node_complex_coherence.py

```diff
@@ -38,322 +38,253 @@
-        """
-        result_size = numpy.prod(self.result_shape(input_shape, max_freq,
-                                                   epoch_length, segment_length,
-                                                   segment_shift, sample_period,
-                                                   zeropad, average_segments)[0]) * 2.0 * 8.0
-        return result_size
-
-
-    def extended_result_size(self, input_shape, max_freq, epoch_length, segment_length,
-                             segment_shift, sample_period, zeropad, average_segments):
-        """
-        Returns the storage size in Bytes of the extended result of the ComplexCoherence. 
-        That is, it includes storage of the evaluated ComplexCoherence attributes
-        such as ...
-        """
-        result_shape = self.result_shape(input_shape, max_freq, epoch_length,
-                                         segment_length, segment_shift,
-                                         sample_period, zeropad, average_segments)[0]
-        result_size = self.result_size(input_shape, max_freq, epoch_length,
-                                       segment_length, segment_shift,
-                                       sample_period, zeropad, average_segments)
-        extend_size = result_size * 2.0  # Main arrays: cross spectrum and complex coherence
-        extend_size = extend_size + result_shape[2] * 8.0  # Frequency
-        extend_size = extend_size + 8.0  # Epoch length
-        extend_size = extend_size + 8.0  # Segment length
-        return extend_size
+                else:
+                    for j in numpy.arange(nepochs):
+                        cs[:, :, f, j] = cs[:, :, f, j] - av[:, f, j] * av[:, f, j].conj().T
+
+    # Compute Complex Coherence
+    ndim = len(cs.shape)
+    if ndim == 3:
+        for i in numpy.arange(cs.shape[2]):
+            temp = numpy.matrix(cs[:, :, i])
+            coh[:, :, i] = cs[:, :, i] / numpy.sqrt(temp.diagonal().conj().T * temp.diagonal())
+
+    elif ndim == 4:
+        for i in numpy.arange(cs.shape[2]):
+            for j in numpy.arange(cs.shape[3]):
+                temp = numpy.matrix(numpy.squeeze(cs[:, :, i, j]))
+                coh[:, :, i, j] = temp / numpy.sqrt(temp.diagonal().conj().T * temp.diagonal().T)
+
+    log.debug("result")
+    log.debug(narray_describe(cs))
+    spectra = spectral.ComplexCoherenceSpectrum(source=time_series,
+                                                array_data=coh,
+                                                cross_spectrum=cs,
+                                                epoch_length=epoch_length,
+                                                segment_length=segment_length,
+                                                windowing_function=window_function)
+    return spectra
+
+
+def complex_coherence_result_shape(input_shape, max_freq, epoch_length, segment_length, segment_shift, sample_period,
+                                   zeropad,
```

Comment:
We could make the extra functions private

Reply:
Ok